File: src/030/p030.ml
open Base
open Stdio

let digits n =
  let rec loop n acc =
    if n = 0 then acc
    else loop (n / 10) (n % 10::acc) in
  match n with
  | 0 -> [0]
  | _ -> loop n []

let sum = List.fold ~init:(0) ~f:(+)

let digits_fifth_power n =
  digits n
  |> List.map ~f:(fun x -> Int.pow x 5)
  |> sum

let solve =
  List.range 2 200000
  |> List.filter ~f:(fun x -> digits_fifth_power x = x)
  |> sum

let () =
  printf "%d\n" solve;;

File: src/030/p030.rs
fn digits(num: i32) -> Vec<i32> {
    num.to_string()
        .chars()
        .map(|c| c.to_digit(10).unwrap() as i32)
        .collect::<Vec<i32>>()
}

fn digits_fifth_power(n: i32) -> i32 {
    digits(n).into_iter()
             .map(|d| d.pow(5))
             .sum()
}

fn solve() -> i32 {
    (2..200000)
        .filter(|&n| digits_fifth_power(n) == n)
        .sum()
}

fn main() {
    let result = solve();
    println!("{}", result);
}
